Eric/add s3 upload task

egieseke opened the pull request:
Adding support for S3 Task.

egieseke committed:
Initial commit for S3 Mule task.

egieseke committed:
Added S3 Tasks and unit tests.

egieseke committed:
Added mule yaml file to test s3 functions.

bricerisingalgorand commented on mule/task/s3/__init__.py:
```diff
@@ -0,0 +1,82 @@
+        super().__init__(args)
+        self.fileName = args['fileName']
+        self.bucketName = args['bucketName']
+        self.objectName = args['objectName'] if 'objectName' in args else None
```
Have you tested setting objectName here? I thought the conditional would need to be` 'objectName' in args.keys()`

egieseke replied:
both ways work: 
>>> args = {'a':1,'b':2,'c':3}
>>> print('found' if 'a' in args else 'not found')
found
>>> print('found' if 'd' in args else 'not found')
not found
>>> print('found' if 'a' in args.keys() else 'not found')
found
>>> print('found' if 'd' in args.keys() else 'not found')
not found

bricerisingalgorand commented:
New dependencies need to be added to setup.py for the pip packages to work

bricerisingalgorand commented on mule/util/s3_util.py:
```diff
@@ -0,0 +1,174 @@
+        this suffix (optional).
+    """
+    s3 = boto3.client("s3")
+    paginator = s3.get_paginator("list_objects_v2")
```
Is there a filter in the s3 library that we could use to reduce the number of objects that are returned?

egieseke replied:
The prefix is added to the kwargs to filter on AWS side, the suffix is used to filter further on the client.

egieseke committed:
Added task DownloadFiles.
Added boto3 dependency to setup.py.
Updated tests not to use algorand-releases s3 bucket.

egieseke committed:
Added new dependencies to setup.py.

egieseke commented:
@bricerisingalgorand , Added new dependencies to setup.py file, including boto3 and botocore.  Please review to sure make they are correct.

btoll commented on mule/task/s3/__init__.py:
```diff
@@ -0,0 +1,103 @@
+        super().execute(job_context)
+        self.download_files()
+
+
```
What about the duplication here?  I know it was mentioned in the design, can we not just have one download files class?

egieseke replied:
The semantics are slightly different, Download file downloads an individual object by objectName and allows renaming the downloaded file.

btoll commented on mule/util/s3_util.py:
```diff
@@ -0,0 +1,174 @@
+    result = True
+    for key in get_matching_s3_keys(bucket_name, prefix, suffix):
+        result &= download_file(bucket_name, key, output_dir)
+    return result
```
Same comment as above.  Can we avoid some code duplication here?

egieseke replied:
Same as above, the semantics are different.

bricerisingalgorand commented:
I got an error from the new logging dependency when I tried to install it
```
➜  mule git:(eric/add-s3-upload-task) python3 -m venv venv
➜  mule git:(eric/add-s3-upload-task) venv/bin/pip3 install -e .
Obtaining file:///Users/bricerising/Documents/git/algorand/mule
Collecting pystache (from mulecli==0.0.0)
Collecting gitpython (from mulecli==0.0.0)
  Downloading https://files.pythonhosted.org/packages/19/1a/0df85d2bddbca33665d2148173d3281b290ac054b5f50163ea735740ac7b/GitPython-3.1.1-py3-none-any.whl (450kB)
     |████████████████████████████████| 460kB 3.8MB/s 
Collecting pygithub (from mulecli==0.0.0)
  Using cached https://files.pythonhosted.org/packages/eb/b6/51eb2ce622632dc7e5be86a7cf47b38a5a25fb837709445f506e8acd10f3/PyGithub-1.47-py3-none-any.whl
Collecting pyyaml (from mulecli==0.0.0)
  Downloading https://files.pythonhosted.org/packages/64/c2/b80047c7ac2478f9501676c988a5411ed5572f35d1beff9cae07d321512c/PyYAML-5.3.1.tar.gz (269kB)
     |████████████████████████████████| 276kB 6.8MB/s 
Collecting termcolor (from mulecli==0.0.0)
Collecting boto3 (from mulecli==0.0.0)
  Downloading https://files.pythonhosted.org/packages/27/87/de75e5a24584d82cca60b86f95d06e56412ed9e23807dcf23896f206f58e/boto3-1.12.39-py2.py3-none-any.whl (128kB)
     |████████████████████████████████| 133kB 8.7MB/s 
Collecting botocore (from mulecli==0.0.0)
  Downloading https://files.pythonhosted.org/packages/dc/fb/f78a0e09965c156fea9160713705af688ec4f18af4249b3095949c930f77/botocore-1.15.39-py2.py3-none-any.whl (6.1MB)
     |████████████████████████████████| 6.1MB 10.4MB/s 
Collecting logging (from mulecli==0.0.0)
  Downloading https://files.pythonhosted.org/packages/93/4b/979db9e44be09f71e85c9c8cfc42f258adfb7d93ce01deed2788b2948919/logging-0.4.9.6.tar.gz (96kB)
     |████████████████████████████████| 102kB 955kB/s 
    ERROR: Command errored out with exit status 1:
     command: /Users/bricerising/Documents/git/algorand/mule/venv/bin/python3 -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/private/var/folders/lp/58jsbh2x23l2kjsmqh8xh2vr0000gp/T/pip-install-8_wy9erg/logging/setup.py'"'"'; __file__='"'"'/private/var/folders/lp/58jsbh2x23l2kjsmqh8xh2vr0000gp/T/pip-install-8_wy9erg/logging/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base pip-egg-info
         cwd: /private/var/folders/lp/58jsbh2x23l2kjsmqh8xh2vr0000gp/T/pip-install-8_wy9erg/logging/
    Complete output (48 lines):
    running egg_info
    creating pip-egg-info/logging.egg-info
    writing pip-egg-info/logging.egg-info/PKG-INFO
    writing dependency_links to pip-egg-info/logging.egg-info/dependency_links.txt
    writing top-level names to pip-egg-info/logging.egg-info/top_level.txt
    writing manifest file 'pip-egg-info/logging.egg-info/SOURCES.txt'
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/private/var/folders/lp/58jsbh2x23l2kjsmqh8xh2vr0000gp/T/pip-install-8_wy9erg/logging/setup.py", line 13, in <module>
        packages = ["logging"],
      File "/usr/local/Cellar/python/3.7.6_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/distutils/core.py", line 148, in setup
        dist.run_commands()
      File "/usr/local/Cellar/python/3.7.6_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/distutils/dist.py", line 966, in run_commands
        self.run_command(cmd)
      File "/usr/local/Cellar/python/3.7.6_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/distutils/dist.py", line 985, in run_command
        cmd_obj.run()
      File "/Users/bricerising/Documents/git/algorand/mule/venv/lib/python3.7/site-packages/setuptools/command/egg_info.py", line 296, in run
        self.find_sources()
      File "/Users/bricerising/Documents/git/algorand/mule/venv/lib/python3.7/site-packages/setuptools/command/egg_info.py", line 303, in find_sources
        mm.run()
      File "/Users/bricerising/Documents/git/algorand/mule/venv/lib/python3.7/site-packages/setuptools/command/egg_info.py", line 534, in run
        self.add_defaults()
      File "/Users/bricerising/Documents/git/algorand/mule/venv/lib/python3.7/site-packages/setuptools/command/egg_info.py", line 570, in add_defaults
        sdist.add_defaults(self)
      File "/usr/local/Cellar/python/3.7.6_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/distutils/command/sdist.py", line 226, in add_defaults
        self._add_defaults_python()
      File "/Users/bricerising/Documents/git/algorand/mule/venv/lib/python3.7/site-packages/setuptools/command/sdist.py", line 127, in _add_defaults_python
        build_py = self.get_finalized_command('build_py')
      File "/usr/local/Cellar/python/3.7.6_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/distutils/cmd.py", line 298, in get_finalized_command
        cmd_obj = self.distribution.get_command_obj(command, create)
      File "/usr/local/Cellar/python/3.7.6_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/distutils/dist.py", line 857, in get_command_obj
        klass = self.get_command_class(command)
      File "/Users/bricerising/Documents/git/algorand/mule/venv/lib/python3.7/site-packages/setuptools/dist.py", line 840, in get_command_class
        self.cmdclass[command] = cmdclass = ep.load()
      File "/Users/bricerising/Documents/git/algorand/mule/venv/lib/python3.7/site-packages/pkg_resources/__init__.py", line 2443, in load
        return self.resolve()
      File "/Users/bricerising/Documents/git/algorand/mule/venv/lib/python3.7/site-packages/pkg_resources/__init__.py", line 2449, in resolve
        module = __import__(self.module_name, fromlist=['__name__'], level=0)
      File "/Users/bricerising/Documents/git/algorand/mule/venv/lib/python3.7/site-packages/setuptools/command/build_py.py", line 15, in <module>
        from setuptools.lib2to3_ex import Mixin2to3
      File "/Users/bricerising/Documents/git/algorand/mule/venv/lib/python3.7/site-packages/setuptools/lib2to3_ex.py", line 12, in <module>
        from lib2to3.refactor import RefactoringTool, get_fixers_from_package
      File "/usr/local/Cellar/python/3.7.6_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/lib2to3/refactor.py", line 19, in <module>
        import logging
      File "/private/var/folders/lp/58jsbh2x23l2kjsmqh8xh2vr0000gp/T/pip-install-8_wy9erg/logging/logging/__init__.py", line 618
        raise NotImplementedError, 'emit must be implemented '\
                                 ^
    SyntaxError: invalid syntax
    ----------------------------------------
ERROR: Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
WARNING: You are using pip version 19.2.3, however version 20.0.2 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
```

bricerisingalgorand reviewed (changes requested):
Need to fix the issue with the logging dependency in the pip install

egieseke committed:
Removed unnecessary dependencies from setup.py.

egieseke commented:
> New dependencies need to be added to setup.py for the pip packages to work



> Need to fix the issue with the logging dependency in the pip install

@bricerisingalgorand cleaned up dependencies so that pip3 -e works.

bricerisingalgorand reviewed (approved):
looks good to me!
